### PR TITLE
feat(apm): add prometheus metrics support

### PR DESCRIPTION
**Pull request by zhengjianhui**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Which docs this PR relation**:

Ref #

**Which toolchain this PR adaption**:

No Breaking Change

**Special notes for your reviewer**:

### CODE REVIEW

File: common/apm/src/prom_server.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+use axum::Router;
+
+pub fn prom_server() -> Router {
+    axum::Router::new().route("/metrics", axum::routing::get(get_metrics))
```
Comment (KaoImin) [inline common/apm/src/prom_server.rs]:
```suggestion
    Router::new().route("/metrics", axum::routing::get(get_metrics))
```

File: common/apm/src/lib.rs [hunk 2]
```diff
@@ -2,9 +2,9 @@
 #![allow(clippy::float_cmp)]
 
 pub mod metrics;
+pub mod prom_server;
 
 pub use muta_apm;
```
Comment (KaoImin) [inline common/apm/src/lib.rs:6]: Rename the module and delete the blank line above.

File: common/apm/src/metrics/api.rs [hunk 3]
```diff
@@ -1,3 +1,4 @@
+#![allow(non_snake_case)]
```
Comment (KaoImin) [inline common/apm/src/metrics/api.rs:1]: Add a blank line below.

File: common/apm/src/prom_server.rs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+use axum::Router;
+
+pub fn prom_server() -> Router {
+    Router::new().route("/metrics", axum::routing::get(get_metrics))
```
Comment (KaoImin) [inline common/apm/src/prom_server.rs]:
```suggestion
    Router::new().route("/metrics", routing::get(get_metrics))
```

File: common/apm/src/prom_server.rs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+async fn get_metrics() -> String {
+    let metrics_data = match super::metrics::all_metrics() {
```
Comment (KaoImin) [inline common/apm/src/prom_server.rs]: Call this function from `crate` rather than from `super`.

File: core/run/src/lib.rs [hunk 6]
```diff
@@ -503,6 +503,15 @@ impl Axon {
         ));
         let _handles = run_jsonrpc_server(self.config.rpc.clone(), api_adapter).await?;
 
+        // Start http server
```
Comment (KaoImin) [inline core/run/src/lib.rs]:
```suggestion
        // Start prometheus http server
```

File: core/run/src/lib.rs [hunk 7]
```diff
@@ -503,6 +503,15 @@ impl Axon {
...
+        tokio::spawn(async move {
+            let addr = std::net::SocketAddr::from(([127, 0, 0, 1], 8100));
```
Comment (KaoImin) [inline core/run/src/lib.rs]: Read the address from the config file.

File: core/run/src/lib.rs [hunk 8]
```diff
@@ -503,6 +503,15 @@ impl Axon {
...
+            axum::Server::bind(&addr)
+                .serve(common_apm::prom_server::prom_server().into_make_service())
+                .await
+                .unwrap();
```
Comment (KaoImin) [inline core/run/src/lib.rs]: Extract into a `run_prometheus_server(addr)` function.

**Review by KaoImin — CHANGES_REQUESTED** [review]

File: common/apm/src/server.rs [hunk 9]
```diff
@@ -1,7 +1,12 @@
 use axum::Router;
+use std::net::SocketAddr;
```
Comment (KaoImin) [inline common/apm/src/server.rs:2]: See this lint https://github.com/cryptape/axon-internal/wiki/Axon-Develop-Lint#order-of-import

File: core/run/src/lib.rs [hunk 10]
```diff
@@ -581,10 +581,7 @@ impl Axon {
             None => std::net::SocketAddr::from(([0, 0, 0, 0], 8100)),
         };
         tokio::spawn(async move {
-            axum::Server::bind(&prometheus_listening_address)
-                .serve(common_apm::server::prometheus_server().into_make_service())
-                .await
-                .unwrap();
+            common_apm::server::run_prometheus_server(prometheus_listening_address);
```
Comment (KaoImin) [inline core/run/src/lib.rs]: use common_apm::server::run_prometheus_server;

**Review by KaoImin — APPROVED** [review]

**Review by driftluo — APPROVED** [review]

**Pull request merged**